src/modules/bulletproofs/main_impl.h

```diff
@@ -0,0 +1,228 @@
+/**********************************************************************
+ * Copyright (c) 2018 Andrew Poelstra                                 *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef SECP256K1_MODULE_BULLETPROOF_MAIN_IMPL
+#define SECP256K1_MODULE_BULLETPROOF_MAIN_IMPL
+
+#include "group.h"
+#include "scalar.h"
+
+#include "modules/commitment/main_impl.h"
+
+struct secp256k1_bulletproof_generators {
+    size_t n;
+    secp256k1_ge *gens;
+    secp256k1_ge *blinding_gen;
+};
```

Comment:
An explanation of what is stored at what array index would have helped me a lot, including a comment that says that `blinding_gen` points into `gens[]`